IntelliTect.TestTools.WindowsTestWrapper/IntelliTect.TestTools.WindowsTestWrapper/DesktopControlHelpers.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Linq;
+using Microsoft.VisualStudio.TestTools.UITesting;
+using Microsoft.VisualStudio.TestTools.UITesting.WpfControls;
+
+namespace IntelliTect.TestTools.WindowsTestWrapper
+{
+    public partial class DesktopControls
+    {
+        public enum SearchTypes
+        {
+            AutomationId,
+            ControlName
+        }
+
+        /// <summary>
+        /// Builds a heirarchy of generic controls by control name.
+        /// </summary>
+        /// <param name="searchType">Searcy by WPF Automation ID or control name</param>
+        /// <param name="baseControl">The first control that needs to be found in the heirarchy</param>
+        /// <param name="controlList">Subsequent controls to find in the heirarchy, ending with the specific control that needs to be acted on (leave blank if the control that 
+        /// needs to be acted on is high enough in the heirarchy that it can be used as the baseControl argument</param>
+        /// <returns>The last control listed if found</returns>
+        protected UITestControl BuildControlHeirarchy(SearchTypes searchType, string baseControl, params string[] controlList)
+        {
+            if (searchType == null)
+                throw new ArgumentNullException(nameof(searchType));
+            if (baseControl == null)
+                throw new ArgumentNullException(nameof(baseControl));
+
+            UITestControl foundControl = new UITestControl();
+
+            if (searchType == SearchTypes.AutomationId)
+            {
+                foundControl = FindWpfControlByAutomationId(baseControl, c => new WpfControl(c), FindWinWindowUnderTest());
+
+                if ( controlList?.Any() != true )
+                {
+                    return foundControl;
+                }
+
+                if ( controlList.Any() )
+                {
+                    foreach (string control in controlList)
+                    {
+                        foundControl = FindWpfControlByAutomationId(control, c => new WpfControl(c), foundControl);
+                    }
+                }
+            }
+
+            if (searchType == SearchTypes.ControlName)
+            {
+                foundControl = FindControlByName(baseControl, c => new UITestControl(c), FindWinWindowUnderTest());
+
+                if (controlList?.Any() != true)
+                {
+                    return foundControl;
+                }
+
+                if ( controlList.Any() )
+                {
+                    foreach (string control in controlList)
+                    {
+                        foundControl = FindControlByName(control, c => new UITestControl(c), foundControl);
+                    }
+                }
+            }
+
+            return foundControl;
+        }
+    }
+}
```

Comment:
There is a bunch of code duplication between these two cases. I think you want something like this:
```C#
protected UITestControl BuildControlHeirarchy(SearchTypes searchType, string baseControl, params string[] controlList)        
{ 
    if (searchType == null) 
        throw new ArgumentNullException(nameof(searchType)); 
    if (baseControl == null) 
        throw new ArgumentNullException(nameof(baseControl)); 
    UITestControl foundControl; 
    switch(searchType)
    {
        case SearchTypes.AutomationId:
            foundControl = FindControlImpl((controlName, rootControl) => FindWpfControlByAutomationId(controlName, c => new WpfControl(c), rootControl)); 
        case SearchTypes.ControlName:
            foundControl = FindControlImpl((controlName, rootControl) => FindControlByName(controlName, c => new UITestControl(c), rootControl));
        default:
            foundControl = new UITestControl();
    }
    return foundControl; 
}

private UITestControl FindControlImpl(Func<string, UITestControl, UITestControl> findMethod, string baseControl, string[] childControls)
{
    UITestControl foundControl = findMethod(baseControl, FindWinWindowUnderTest()); 
    if ( childControls?.Any() == true) 
    { 
        foreach (string control in controlList) 
        { 
            foundControl = findMethod(control, foundControl);
        } 
    } 
    return foundControl;
}
```

Few things to note:
- `foundControl` starts as `null` and we only give it the default value in the unlikely case where we have not set it to anything else
- Use a switch rather than a bunch of if statements when acting on an enum.
- The common logic is abstracted into its own method
- Added null propogation operator when checking childControls. Typically you don't need to check values that come from params as the compiler will put in an empty array unless the caller specifically passes null.
- Zero testing of this code on my part, so there may be errors.